### ISSUE TITLE
Do not add chunked Transfer-Encoding header to 304 response when upstream doesn't provide Content-Length

### ISSUE THREAD

**Issue by boringuy** [envoyproxy/envoy#16071]
*Title*: *Do not add chunked Transfer-Encoding header to 304 response when upstream doesn't provide Content-Length*

*Description*:
This is not exactly a bug according to the spec:

> Transfer-Encoding MAY be sent in a response to a HEAD request or in a
   304 (Not Modified) response (Section 4.1 of [RFC7232]) to a GET
   request, neither of which includes a message body, to indicate that
   the origin server would have applied a transfer coding to the message
   body if the request had been an unconditional GET.  This indication
   is not required, however, because any recipient on the response chain
   (including the origin server) can remove transfer codings when they
   are not needed.

However, in our use case, we cannot be sure if adding Transfer-Encoding to a 304 response would affect the client. Since 304 response (similar to response to HEAD request) should NEVER have a body. Content-Length and Transfer-Encoding header should be the indication of the Headers you would get (minus the body) if you do unconditional GET. For a 304 response, Content-Length and Transfer-Encoding are optional, so, it's ok for upstream to not set it. I think it would be best behavior and less confusing if envoy would just do what upstream does on 304 response and proxy it through. There is really no benefit at all to add Transfer-Encoding chunked to a 304 response. 

[optional *Relevant Links*:]
https://tools.ietf.org/html/rfc7230#section-3.3.1

**Comment by asraa** [comment]
I think that makes sense to me. An option would definitely be good, with default preserving behavior for now.
I would tag Alyssa, but she is out -- @antoniovicente @yanavlasov other opinions?

**Comment by antoniovicente** [comment]
Seems about right.  304 responses shouldn't have a body; handling should be similar to 204 responses.

**Comment by boringuy** [comment]
While working on the integration test, I found that envoy is also adding "content-length: 0" to 304 response when upstream doesn't supply content-length header. This is definitely a bug as the "content-length" in a 304 response is NOT the content-length of the 304 response but should be the content-length of the original assert if it were to return 200. 

The fix that I am doing actually also fixes this. So, I will opt for doing this without an option guarding it first and get input from the PR to see if we need an option for this and if I am fixes this the right way. 

Basically, what I am doing is in codec_impl.cc:
```
-    if (status && *status == 100) {
+   if (status && (*status == 100 || *status == 304)) {
      // Make sure we don't serialize chunk information with 100-Continue headers.
+      // For 304 response, since it should never have a body, we should not need to chunk_encode at all.
      chunk_encoding_ = false;
```

**Comment by boringuy** [comment]
pull request submitted but need approval as this is my first PR: https://github.com/envoyproxy/envoy/pull/16545
